File: Python/shortest_path/dijkstra.py
import heapq
from heapq import heappush, heappop
from typing import List, Tuple
from math import inf

"""
Dijkstra算法
有向图D, 根结点root
1. 初始化 l(x) - root -> x 最短路, l[x] = inf, l[root] = 0;
   visited数组; q 中存储下一个起点
2. 更新与 q 中存储的当前节点 v 存在 v->x 有向路的 l[x] = min(l[x], l[v]+g[v][x])
3. 重复步骤2, 直到 q 为空
"""
def dijkstra_dense(root: int, g: List[List[int]]):
    n = len(g)
    l = [inf] * n
    l[root] = 0
    vis = [False] * n
    vis[root] = True

    q = [root]
    while q:
        u = q.pop()
        nxt, nxt_mi = -1, inf
        for v in range(n):
            if vis[v]:
                continue
            if g[u][v]:
                l[v] = min(l[v], l[u]+g[u][v])
            if l[v] < nxt_mi:
                nxt_mi = l[v]
                nxt = v
        if nxt != -1:
            q = [nxt]
            vis[nxt] = True
    return l

# g 是稀疏临界图
def dijkstra_sparse(root: int, g: List[List[Tuple(int, int)]]):
    n = len(g)
    d = [inf] * n
    pq = [(0, root)]
    while pq:
        du, u = heappop(pq)
        if du > d[u]: # u is already updated
            continue
        for v, w in g[u]:
            nd = du + w
            if nd < d[v]:
                d[v] = nd
                heappush(pq, (nd, v))
    return d

if __name__ == "__main__": 
    edges = [
                [0, 1, 7],
                [0, 2, 1],
                [1, 3, 4],
               # [3, 1, 4],
                [1, 5, 1],
                [2, 1, 5],
                [2, 4, 2],
                [2, 5, 7],
                [4, 1, 2],
                [4, 3, 5],
               # [3, 4, 5],
                [5, 4, 3]
            ]
    n = 6
    g = [[0 for _ in range(n)] for _ in range(n)]
    for e in edges:
        g[e[0]][e[1]] = e[2]
        g[e[1]][e[0]] = e[2] # 无向图

    r = input()
    while r:
        print(dijkstra_dense(int(r), g))
        r = input()
    

    g = [[] for _ in range(n)]
    for u, v, w in edges:
        g[u].append((v, w))
        g[v].append((u, w))


    r = input()
    while r:
        print(dijkstra_sparse(int(r), g))
        r = input()

File: C++/shortest_path/dijkstra.cpp
#include <functional>
#include <iostream>
#include <list>
#include <queue>
#include <utility>
#include <vector>
#include <climits>

using namespace std;

ostream& operator<<(ostream& os, const vector<int>& v){
  for (auto& i: v){
    os << i << " "; 
  }
  os << endl;
  return os;
}

// priority_queue version
class Dijkstra 
{
private:
  int m_numNodes;
  vector<vector<int>> m_edges;
  vector<vector<pair<int,int>>> m_graph;
  vector<vector<int>> shortest_path;
public:
  explicit Dijkstra(int n_nodes, const vector<vector<int>>& edges) 
    : m_numNodes(n_nodes), m_edges(edges) 
  {
    m_graph.resize(m_numNodes);
    shortest_path.resize(m_numNodes);
    for (const auto& e : m_edges){
      int vert_start = e[0], vert_end = e[1], dist = e[2];
      m_graph[vert_start].emplace_back(vert_end, dist);
    }
  };

  vector<int> solve(int root){
    if (!shortest_path[root].empty()) return shortest_path[root];
    vector<int> ans(m_numNodes, INT_MAX);
    ans[root] = 0;

    priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> pq;
    pq.emplace(0, root);
    while (!pq.empty()){
      auto [d, cur] = pq.top();
      pq.pop();
      if (d > ans[cur]) continue;
      for (const auto& [nxt, d2nxt]: m_graph[cur]){
        if (ans[nxt] > ans[cur] + d2nxt){
          ans[nxt] = ans[cur] + d2nxt;
          pq.emplace(ans[nxt], nxt);
        }
      }
    }
    shortest_path[root] = ans;
    return ans;
  }

};

vector<int> dijkstra(int root, const vector<vector<int>>& g){
  int n = g.size();
  vector<int> l(n, INT_MAX);
  if (root < 0 || root >= n) return l;

  vector<int> vis(n, false);
  l[root] = 0;
  vis[root] = true;
  
  vector<int> q;
  q.push_back(root);
  while (!q.empty()){
    int v = q[0];
    q.pop_back();

    // next vertex
    int cur = INT_MAX, cur_i = -1;
    for (int i = 0; i < n; i++){
      if (!vis[i] && g[v][i]){
        l[i] = min(l[i], l[v]+g[v][i]);
        if (cur > l[i]){
          cur = l[i];
          cur_i = i;
        }
      }
    }
    if (cur_i != -1){
      q.push_back(cur_i);
      vis[cur_i] = true;
    }
  }

  return l;
}


int main(){
  int n = 6;
  vector<vector<int>> edges{
    {0, 1, 7},
    {0, 2, 1},
    {1, 3, 4},
    {1, 5, 1},
    {2, 1, 5},
    {2, 4, 2},
    {2, 5, 7},
    {4, 1, 2},
    {4, 3, 5},
    {5, 4, 3}
  };

  Dijkstra shortest_path_solver = Dijkstra(n, edges);
  cout << shortest_path_solver.solve(0);


  vector<vector<int>> g(n, vector<int>(n));

  for (auto& e: edges){
    g[e[0]][e[1]] = e[2];
  }

  cout << dijkstra(0, g);
}
